sample/weighted.go

```diff
@@ -0,0 +1,104 @@
+// Copyright ©2015 The gonum Authors. All rights reserved.
+// Use of this code is governed by a BSD-style
+// license that can be found in the LICENSE file
+
+package sample
+
+import "math/rand"
+
+// Weighted is a collection of items that can be selected with weighted probabilities
+// without replacement.
+type Weighted struct {
+	indexes []int
+	weights []float64
+	totals  []float64
+	src     *rand.Rand
+}
+
+// NewWeighted returns a Weighted for the indexes in idx weighted by w. Elements of
+// w are modified during sampling, so a copy should be passed if the user needs to
+// keep the values. NewSampler panics if len(idx) != len(w). If src is nil,
+// rand.Rand is used as the random source.
+func NewWeighted(idx []int, w []float64, src *rand.Rand) Weighted {
+	if len(idx) != len(w) {
+		panic("floats: length of the slices do not match")
+	}
+	s := Weighted{
+		indexes: idx,
+		weights: w,
+		totals:  make([]float64, len(w)),
+	}
+	s.reset()
+	return s
+}
+
+func (s Weighted) reset() {
+	copy(s.totals, s.weights)
+	for i := len(s.totals) - 1; i > 0; i-- {
+		// Sometimes 1-based counting makes sense.
+		s.totals[((i+1)>>1)-1] += s.totals[i]
```

Comment:
Time complexity; one is O(m + nlogm), the other is O(nm) where m is size of the collection and n is the number of samples.